### ISSUE TITLE
Does this work?

### ISSUE THREAD

**Issue by angusturner** [revsic/jax-variational-diffwave#1]
Hi, I have been trying to implement VDM as well, and am just wondering whether you have had any success with this?

I am working in PyTorch with the discrete-time version, but I am having trouble with numerical stability. It is not directly comparable, but most of our code seems to match up.

One difference I notice is that you constrained the biases of logSNR function to be positive - I *think* you can get away with positive weights and unconstrained biases (to satisfy monotonic constraint), but I'm willing to be corrected here. (Not sure if this makes much of a difference).

Otherwise, this is a really awesome implementation! I'm just curious if it works because I am struggling to make my own version work!

Cheers

**Comment by revsic** [comment]
Hi, Turner. Thank you for your participation.

First, I fail my all experiments, and I'm estimating the reason that the expressivity of wavenet is not enough to model the continuous-time loss. So I try to alter the model with HiFi-GAN generator nowadays.

Second, if the `bias` you refer was the variable `gamma_gap`, I think it should be positive to satisfy the monotonic constraint, since the `normalized -logSNR` is in range [0, 1]. Otherwise, if it was the variable `bias` of `PosDense`, I think it should be too, but I can't be sure that I don't have an intuition about that proof.

Precisely, it works and loss converges. But it could not recover the spectrogram, and I thought the loss should be converged at lower points (in the previous study, it converges at 8e-3, [tf-diffwave](https://github.com/revsic/tf-diffwave), but now at 0.16). This was the basis for my intuition, lack of expressivity of the generator model. 

Sorry for my bad English, if I miss understood your questions, please correct it.
Thanks

![image](https://user-images.githubusercontent.com/15906088/134170830-80f97ab8-0c7f-4205-a99c-2c585f233ed6.png)
![image](https://user-images.githubusercontent.com/15906088/134170952-cdf37f5d-071e-4f09-b32c-f254471315c0.png)

**Comment by angusturner** [comment]
Thanks, these training plots are quite helpful!

For the monotonic constraint, I was referring to the `bias` of `PosDense` - I think this should be unconstrained. `PosDense` will be monotonic as long as it has a positive derivative with respect to the input `timestep`. The bias terms do not affect this derivative. (I am not 100% sure, but I have also seen other examples doing unconstrained bias, such as this [Neural Autoregressive Flow](https://github.com/CW-Huang/torchkit/blob/master/torchkit/flows.py#L281) implementation.) Maybe this is important for the flexibility of the learned `logSNR` function? 

The other question I have is about the calculation of the loss terms? For example, with the prior and reconstruction, I wonder what happens if you use the NLL and KL instead of MSE? For example, maybe this could affect the relatively scales of the different loss terms?

For example, if you do something like`rec_loss = -Normal(q_mu_z0, q_logvar_z0).log_pdf(x)` ? The scale will be quite different to using `MSE`, which might affect learning? (And likewise, for prior loss compute analytic KL from `q(z1|x)` and `N(0, 1)` ).

Anyway, these are just some ideas. Interested to follow along if you get things working!

**Comment by revsic** [comment]
I agree with your opinion that constrained bias can hurt the expressivity of `logSNR`, since model use `sigmoid` activation. I'll test the bias-unconstrained PosDense in near days.
(1. constrained `PosDense` would generate only positive numbers from input time [0, 1], 2. then it will be fed to the `sigmoid`, 3. the output range [0, 1] limits to [0.5, 1]. 4. cannot utilize whole ranges)

The earlier versions of this repo use the NLL and KL, as you referred `-Normal(q_mu_z0, q_logvar_z0).log_pdf(x)`.
Analytically, for `z = alpha x signal + std x noise`, reconstruction loss and posterior loss can be rewritten as `nll(z, mean=signal, stddev=1) \propto (z - signal) ** 2 + C` and `KL(z || N(0, 1)) \propto z ** 2 - 2logstd + C`. In posterior loss, `z ** 2` is derived from N(0, 1) log-likelihood and `2logstd` acts like preventing posterior collapse. But in optimization perspective, square loss tries to reduce both `alpha` and `sigma`, during `2logstd` tries to increase the `sigma`. It causes the training instability in previous experiments, and so after version use `(z - noise) ** 2` to increase only sigma and simply use square loss for reconstruction loss too. 
(`KL(z || N(0, 1)) = nll(z, mean=0, stddev=1) - nll(z, mean=alpha x signal, stddev=std) \propto z ** 2 - 2logstd - (z - alpha x signal) ** 2 / std ** 2 + C = z ** 2 - 2logstd - noise ** 2 + C = z ** 2 - 2logstd + C`)

Of course, this is theoretically incorrect, but in experiments, it helps the convergence of the prior and reconstruction loss. I erased the earlier loss plots before, so I can't show you how this loss affects the training curves. So I think it would be nice if you think of it as just heuristics.

**Comment by revsic** [comment]
Sorry, it was not so affective (red for unconstrained posdense, orange for original one). It could make a small difference, not the solution for the major early-convergence problem, since `err` term starts to converge on the same points. 
(I'll update the loss plots on near futures, more learning epochs)

I'll close this issue. Please reopen the issue if you have another questions.
Thanks
![image](https://user-images.githubusercontent.com/15906088/134816911-38934b4c-3b90-407b-a9e3-190ec2d8de42.png)

**Comment by rardz** [comment]
> I agree with your opinion that constrained bias can hurt the expressivity of `logSNR`, since model use `sigmoid` activation. I'll test the bias-unconstrained PosDense in near days.
> (1. constrained `PosDense` would generate only positive numbers from input time [0, 1], 2. then it will be fed to the `sigmoid`, 3. the output range [0, 1] limits to [0.5, 1]. 4. cannot utilize whole ranges)
> 
> The earlier versions of this repo use the NLL and KL, as you referred `-Normal(q_mu_z0, q_logvar_z0).log_pdf(x)`.
> Analytically, for `z = alpha x signal + std x noise`, reconstruction loss and posterior loss can be rewritten as `nll(z, mean=signal, stddev=1) \propto (z - signal) ** 2 + C` and `KL(z || N(0, 1)) \propto z ** 2 - 2logstd + C`. In posterior loss, `z ** 2` is derived from N(0, 1) log-likelihood and `2logstd` acts like preventing posterior collapse. But in optimization perspective, square loss tries to reduce both `alpha` and `sigma`, during `2logstd` tries to increase the `sigma`. It causes the training instability in previous experiments, and so after version use `(z - noise) ** 2` to increase only sigma and simply use square loss for reconstruction loss too.
> (`KL(z || N(0, 1)) = nll(z, mean=0, stddev=1) - nll(z, mean=alpha x signal, stddev=std) \propto z ** 2 - 2logstd - (z - alpha x signal) ** 2 / std ** 2 + C = z ** 2 - 2logstd - noise ** 2 + C = z ** 2 - 2logstd + C`)
> 
> Of course, this is theoretically incorrect, but in experiments, it helps the convergence of the prior and reconstruction loss. I erased the earlier loss plots before, so I can't show you how this loss affects the training curves. So I think it would be nice if you think of it as just heuristics.

In your brief explanation of the reconstruction loss, which @angusturner referred as `-Normal(q_mu_z0, q_logvar_z0).log_pdf(x)`, you firstly replaced it with: `nll(z, mean=signal, stddev=1)`, why the standard deviation item is default to 1? I can't find any clue of the reconstruction loss in the paper. However in my opinion, `stddev` should be 0 (`q_logvar_z0` to be negative infinity) or a small number that is very close to 0, i.e. from `z0` to `x` should be as deterministic as possible.

**Comment by revsic** [comment]
They represent the reconstruction loss in equation 15 as `E_{q(z_0|x)}[-log p(x|z_0)]`, which forms `-Normal(mu_{x|z_0}, sigma_{x|z_0}).log_pdf(z_0) where z_0 ~ q(z_0|x)`. It is different with `-Normal(q_mu_z0, q_logvar_z0).log_pdf(x)`, which can be rewritten as `E_{p(x)}[-log q(z_0|x)]`. As you @rardz metioned, they just write the reconstruction term without any clue of `p(x_i|z_{0, i})` in equation 9. So I just assume the `p(x_i|z_{0, i})` as simple `N(x_i, 1)`.
If we assume it as `N(x_i, eps)` with small epsilon, in optimization perspective, the sizes of the loss vary in too different levels (i.e. prior loss and diffusion loss live in [0, 1], reconstruction loss lives in maybe [0, eps^{-1}]), and it can cause the optimization instability I think. 

If you want to model the `sigma_{x|z_0}`, which I assume as 1 in `N(x_i, 1)`, I suggest the paper `Improved Denoising Diffusion Probabilistic Models`, arXiv:2102.09672. 
Thanks.

**Comment by rardz** [comment]
1. `E_{q(z_0|x)}[-log p(x|z_0)]` should form `-Normal(mu_{x|z_0}, sigma_{x|z_0}).log_pdf(x)`,
2. as they noted in paper:
> where we choose p(xi|z0,i) \propto q(z0,i|xi). With sufficiently large SNR(0), this becomes a very close approximation to the true q(z0|x), as the influence of the unknown data distribution q(x) is overwhelmed by the likelihood q(z0|x).

it seems that they tend to use the distribution of q(z0|x) for reconstruction estimation, thus `-Normal(mu_{x|z_0}, sigma_{x|z_0}).log_pdf(x) \approx -Normal(q_mu_z0, q_logvar_z0).log_pdf(x)`, which as you @revsic mentioned is `E_{p(x)}[-log q(z_0|x)]`.

So, I think `-Normal(q_mu_z0, q_logvar_z0).log_pdf(x)` should be the author's way of reconstruction loss calculation. Of course it was just my one-sided guess.

**Comment by revsic** [comment]
Right, I misunderstood the formula and agree with the form `-Normal(mu_{x|z0), sigma_{x|z0}).log_pdf(x)`. I will fix it and retest the experiment. 
Thanks for your contribution.

**Comment by rardz** [comment]
Actually I have done a few trials with some toy datasets, and found it hardly to replicate the noise schedule curve (titled by log SNR vs time t in Figure 4)  learned in that VDM's paper. In my case, SNR_max increases as optimization going on and doesn't seem to converge.
However, maybe I was wrong in some hidden corners. Hoping you @revsic would get a good result.

**Comment by yoyolicoris** [comment]
I also did a [re-implementation](https://github.com/yoyololicon/variational-diffwave) of VDM using diffwave and I can share some results I got.

@angusturner I would suggest you check the [open review](https://openreview.net/pdf?id=2LdBqxc1Yv) version of VDM paper. With Eq. 14 you should get rid of the numerical issue.

> So, I think `-Normal(q_mu_z0, q_logvar_z0).log_pdf(x)` should be the author's way of reconstruction loss calculation. Of course it was just my one-sided guess.

@rardz I can confirm this works well in my implementation.

Some plots I got when training with T = 1000:
![image](https://user-images.githubusercontent.com/17811395/152463801-37eedcb8-664d-4a4e-862a-3718f744cf52.png)

The final negative VLB was around ~= -5.75.

The learned noise schedule:
![image](https://user-images.githubusercontent.com/17811395/152463997-1725b394-0977-45aa-8345-9bd1c3d48fbb.png)

Although my implementation can work quite well with discrete-time loss, it failed on training with continuous-time. 
I have tried to disable the variance minimization trick or detach the gradients of the derivative of logSNR when computing loss,  but in both ways, the learned noise schedule curve still bends in the opposite direction. I didn't check very carefully, but most of the codes are matched with @revsic.

**Comment by revsic** [comment]
continous-time modeling is so hard.. thank you for your comment @yoyololicon

**Comment by angusturner** [comment]
Hi, I know it has been a long time, but I have revisited this recently and found that I can get VDM to work with the correct scaling of reconstruction and prior terms.

I will admit, I have not tried it on raw audio, but on toy problems and another task I am trying for work I can get reasonable results.

For the KL / prior term I use the analytic solution from the official VDM code:
https://github.com/google-research/vdm/blob/main/model_vdm.py

For the reconstruction, I have continuous data, and set it up as follows:
```
p(x|z0) = q(z0|x)p(x)/p(z0)
p(x|z0) is proportional to q(z0|x) (with p(x) as unknown constant, p(z0) also unknown), 
```

Therefore, I simply optimize `rec_loss = -Normal(q_mu_z0, q_logvar_z0).log_pdf(z_0)`. (i.e. minimize the entropy of the first forward step).

The loss scale looks weird visually (I presume because we are missing the normalization terms), but the training seems stable and the learned values are sensible (actually quite close to the learned values in Kingma's code).


In the case of discrete data it seems you can actually compute the normalization `p(z_0) = sum_x{q(z|x_0}` for all 256 pixel values. I imagine you could probably make an estimate here with monte carlo sampling too. I'm going to try some more things. But for now this seems to work...

tldr; it kind of works with just`rec_loss = -Normal(q_mu_z0, q_logvar_z0).log_pdf(z_0)`

**Comment by rardz** [comment]
The official repo now updates with a colab example:
https://colab.research.google.com/github/google-research/vdm/blob/main/colab/SimpleDiffusionColab.ipynb

In that case,  their use a simple resnet to predict the output (in their case it's discrete logit).
So this may be a very suitable trick for calculating the final reconstruction loss, by adding some simple layers for the output distribution prediction.
